Use IntersectionObserver if available

snewcomer opened the pull request:
## IntersectionObserver API

This PR provides an easy upgrade path to use IntersectionObserver.  This also doesn't have any effect on existing users and will fallback to rAF if not in target browser.

Also added two routes - `infinity` and `infinity-scrollable` to perf-test and test.

- [x] use IntersectionOverride by default
- [x] allow for scrollable area (a configuration option on the IntersectionObserver)
- [x] tests (including separate tests for IntersectionObserver, rAF and scrollEvent)
- [x] documentation

close #106 #67 
Ref prev PR #112 
Partial Fix #37

snewcomer commented:
So I guess my first question is - to start, should users opt into using the IntersectionObserver API with plans to make it the default down the line?  Perhaps using ember-spaniel once this [PR](https://github.com/linkedin/spaniel/pull/63) is merged?  Or should IntersectionObserver be the default and have the users update their viewport configs?

snewcomer commented on addon/mixins/in-viewport.js:
```diff
@@ -91,22 +103,51 @@ export default Mixin.create({
-        bind(this, this._setViewportEntered, context)
-      );
+    if (get(this, 'viewportUseIntersectionObserver')) {
+      const { top, left, bottom, right } = this.viewportTolerance;
```
Maybe can make this a separate config item specific to IntersectionObserver.

snewcomer commented on tests/dummy/app/controllers/infinity.js:
```diff
@@ -0,0 +1,18 @@
+const models = [...Array(10).fill().map(() => `https://s3.amazonaws.com/uifaces/faces/twitter/${images[(Math.random() * images.length) | 0]}/128.jpg`)];
+
+export default Ember.Controller.extend({
+  models,
```
This route uses optimistic rendering.  Guessing we don't want to control this correct and leave that up to the users?  i.e. 2 requests will be made - first for the 1st set of records.  While that is pending, it will trigger `infinityLoad()`.

snewcomer commented:
Switched this over to use the IntersectionObserver API by default.

snewcomer commented on addon/mixins/in-viewport.js:
```diff
@@ -91,22 +103,51 @@ export default Mixin.create({
+    if (get(this, 'viewportUseIntersectionObserver')) {
+      const { top, left, bottom, right } = this.viewportTolerance;
+      const options = {
+        root: null, 
```
Another question - if null, the root is the layout view (document.documentElement).  One case I have run up against before is the scrollable area is not the window but rather a confined area in the viewport.

Would you be in favor of a `scrollableArea` config variable?

snewcomer commented:
Hey! Sry to be a bug.  A review would be great when (or more apt 'if' :) ) somebody has time!

:eyes:  

@poteto @kellyselden

snewcomer commented on addon/mixins/in-viewport.js:
```diff
@@ -177,22 +219,21 @@ export default Mixin.create({
+  _unbindScrollDirectionListener() {
     const elementId = get(this, 'elementId');
 
+    const context = get(this, 'scrollableArea') ? get(this, 'scrollableArea') : window;
```
Could probably make a variable on this mixin instead of this ternary.

sbatson5 replied:
`const context = get(this, 'scrollableArea') || window` ?

snewcomer commented on tests/dummy/app/routes/infinity-scrollable-scrollevent.js:
```diff
@@ -0,0 +1,4 @@
+import ScrollableRoute from './infinity-scrollable';
+
+export default ScrollableRoute.extend({
```
Don't necessarily like this, but to avoid duplication and allow for deterministic tests, I separated out each InfinityObserver, rAF, and scrolEvent options in the lib to three separate routes with a scrollable area.

snewcomer commented:
Also just to note, this PR would benefit ember-infinity ([PR](https://github.com/hhff/ember-infinity/pull/242)) && ember-light-table ([related](https://github.com/offirgolan/ember-light-table/pull/519)).

BartTK commented:
I've tested this change on my project and it scrolls a lot smoother then the rAF approach. However I was disappointed with respect to the loading time. When having a few thousands components on a page using this mixin, it takes several seconds to load them. In my specific case, all components used the same scrollableArea and the same options. As a result, I could re-use a single IntersectionObserver instance that observes all components. This greatly improved the loading time.

simonihmig commented:
Would love to see this land! 👍
Anything missing here? (except for the rebase)

snewcomer commented:
I'm going to upgrade to 2.18 and use the new testing apis in a separate commit.  Getting a test leaking/flakey test.

snewcomer commented:
So since I hadn't got a review yet, I updated the pkg to 2.18 and changed everything over to the new testing apis and `@ember` module namespace.  Part of it was to fix a flakey test since the qunit container changed from 2.7 to 2.17 and then it snowballed from there 🤦‍♂️   

I know it isn't a very nice thing to do from a reviewers POV, so I can easily break the last two commits into another PR if everybody would like.  However, since most of main changes are new files, upgrading didn't change much:

1.  Add IntersectionObserver API in `in-viewport` mixin
2.  Add acceptance tests (totally new files).
3.  New routes for testing IntersectionObserver, rAF, and scroll events (totally new files)

cibernox commented:
@snewcomer is it possible to submit the update to 2.18 in a separate PR?

I haven't reviewed this because I'm not familiar with the codebase. TBH I don't know who is the best person to review it.

snewcomer commented:
@BartTK great insight btw.  Curious, how did you get one IntersectionObserver to rule them all?  I’m thinking your scenario might be somewhat interesting....where we could also somehow use one IntersectionObserver to ‘observe’ each element instead of many IntersectionObservers to ‘observe’ one element.  Not totally sure of the perf implications, but your observation makes it seem like we should look into this. (after this PR)

BartTK commented:
You only need a single IntersectionObserver per scrollableArea/options pair. When creating the observer, add it to a static administration. Future components having the same scrollableArea and options can re-use that IntersectionObserver instead of creating a new one. Just call 'observe' again for the new component.

There are a few things to keep in mind. You need to do reference counting to know when to destroy the observer. Furthermore you cannot bind '_onIntersection' to 'this' anymore; a single callback will be used by multiple components. Instead of receiving a single entry, this callback now receives a list of entries. These entries only contain the DOM element, you need to maintain another administration to map that element to its component.

In my case, this resulted in a large performance boost with respect to the loading time of the page.

snewcomer commented:
@BartTK cool 😎 so you forked and modified this PR then?  Once this PR gets in, and of course if you have time, it would be a great PR!

sbatson5 commented on README.md:
```diff
@@ -100,11 +103,33 @@ export default Ember.Component.extend(InViewportMixin, {
   Default: Depends on browser
 
-  As it's name suggests, if this is `true`, the Mixin will use `requestAnimationFrame` instead of the Ember run loop. Unless you want to force enabling or disabling this, you won't need to override this option.
+  As it's name suggests, if this is `true` and the IntersectionObserver API is not available in the target browser, the Mixin will use `requestAnimationFrame`. Unless you want to force enabling or disabling this, you won't need to override this option.
```
`its` instead of `it's`

sbatson5 commented on tests/acceptance/infinity-test.js:
```diff
@@ -0,0 +1,63 @@
+
+  andThen(() => {
+    assert.equal(findAll('.infinity-svg-scrollEvent').length, 20);
+    // assert.equal(find('.infinity-scrollable-scrollEvent.inactive').length, 1, 'component is inactive after fetching more data');
```
Commented out test

sbatson5 commented on tests/dummy/app/routes/infinity-scrollable.js:
```diff
@@ -0,0 +1,17 @@
+    return [...Array(10).fill().map(() => `${images[(Math.random() * images.length) | 0]}`)];
+  },
+  setupController(controller, model) {
+    set(controller, 'model', model);
```
Isn't this the default behavior of `setupController`?  Is there a reason to overwrite?

sbatson5 commented on tests/dummy/app/templates/infinity-scrollable.hbs:
```diff
@@ -0,0 +1,15 @@
+<div class="list" style="height: 500px; overflow: scroll;">
+  <ul>
+    {{#each model as |val|}}
+    <div class="infinity-svg">
```
Indent?  Only to be consistent with the next file where the `each` block is indented

Duder-onomy commented:
BOOM!
